src/librbd/api/Namespace.cc

```diff
@@ -64,22 +64,44 @@ int Namespace<I>::create(librados::IoCtx& io_ctx, const std::string& name)
     return r;
   }
 
+  int ret_val;
+  bool remove_dir_obj = false;
   librados::IoCtx ns_ctx;
   ns_ctx.dup(io_ctx);
   ns_ctx.set_namespace(name);
 
+  r = ns_ctx.create(RBD_TRASH, false);
+  if (r < 0) {
+    lderr(cct) << "failed to create trash: " << cpp_strerror(r)
+               << dendl;
+    goto rollback;
+  }
+
   r = cls_client::dir_state_set(&ns_ctx, RBD_DIRECTORY,
                                 cls::rbd::DIRECTORY_STATE_READY);
   if (r < 0) {
     lderr(cct) << "failed to initialize image directory: " << cpp_strerror(r)
                << dendl;
+    remove_dir_obj = true;
     goto rollback;
   }
 
   return 0;
 
 rollback:
-  int ret_val = cls_client::namespace_remove(&default_ns_ctx, name);
+  if (remove_dir_obj) {
+    ret_val = ns_ctx.remove(RBD_DIRECTORY);
+    if (ret_val < 0 && ret_val != -ENOENT) {
+      lderr(cct) << "failed to remove image directory: " << cpp_strerror(ret_val) << dendl;
+    }
+  }
+
+  ret_val = ns_ctx.remove(RBD_TRASH);
```

Comment:
How about replacing remove_dir_obj with remove_trash_obj and setting that to true only if the trash creation succeeds?  It would be neater than multiple gotos.

Reply:
I don't see anything wrong with multiple gotos.  If named appropriately, they are harder to screw up than something based on dummy variables.